### ISSUE TITLE
Refresh tokens that never expire?

### ISSUE THREAD

**Issue by kentcb** [IdentityServer/IdentityServer4#4705]
### Question

Is there a canonical way of setting up refresh tokens that never expire? For context, it's for scenarios where we have devices registered to use our APIs. We control the devices and the user credentials used to register those devices. We also control revocation of those accounts and associated tokens. Once a device is registered, we want it to be able to renew access tokens indefinitely without any front channel communications (it's a kiosk device).

Based on reading [these docs](https://identityserver4.readthedocs.io/en/latest/topics/refresh_tokens.html), I've currently got this configuration:

```C#
AbsoluteRefreshTokenLifetime = 0,
RefreshTokenExpiration = TokenExpiration.Sliding,
SlidingRefreshTokenLifetime = int.MaxValue,
```

If I do this with a super short access token expiration and exercise the client, it _seems_ to work fine sometimes but then other times it will pop open the browser momentarily (momentarily because the browser has a cached cookie and the user doesn't need to re-auth). When this occurs, I see this in the IS4 log:

```
warn: IdentityServer4.Validation.TokenValidator[0]
      Refresh token has expired. Removing from store.
warn: IdentityServer4.Validation.TokenRequestValidator[0]
      Refresh token validation failed. aborting, {
        "ClientId": "deviceregistrations",
        "ClientName": "Device Registrations",
        "GrantType": "refresh_token",
        "Raw": {
          "grant_type": "refresh_token",
          "refresh_token": "***REDACTED***",
          "client_id": "deviceregistrations",
          "client_secret": "***REDACTED***"
        }
      }
```
But based on my configuration I don't understand _how_ the refresh token could have expired. I've checked our backing store for grants (Firestore) and it looks fine:

![image](https://user-images.githubusercontent.com/1901832/89161128-8aa3aa00-d5b5-11ea-8a49-9e5160ef5604.png)

Moveover, I've debugged through our `IPersistedGrantStore` implementation and made sure that the `PersistedGrant` it is returning looks correct:

![image](https://user-images.githubusercontent.com/1901832/89161665-41a02580-d5b6-11ea-8d8b-3e00b446cce1.png)

The only other clue I have right now is this warning appearing in the log:

```
warn: IdentityServer4.Endpoints.TokenEndpoint[0]
      Invalid HTTP request for token endpoint
```

I assume `TokenEndpoint[0]` refers to the first client in my list, which is indeed `deviceregistrations`. I'm not sure if this warning is relevant given that I see it both when the token renewal succeeds and when it fails due to expiration.

Can anyone offer any insight as to why the refresh token might be failing to validate?

**Comment by kentcb** [comment]
Spelunking my way through the code, it seems to be hitting this case in `DefaultRefreshTokenService.cs`:

```C#
/////////////////////////////////////////////
// check if refresh token has expired
/////////////////////////////////////////////
if (refreshToken.CreationTime.HasExceeded(refreshToken.Lifetime, Clock.UtcNow.DateTime))
{
    Logger.LogWarning("Refresh token has expired.");
    return invalidGrant;
}
```

Unfortunately, I'm not set up to debug this yet (seems like it's gonna be quite an exercise to do so), so I've tried to debug it in my head and it's not making a lot of sense why this would trigger.

**Comment by kentcb** [comment]
So I've managed to debug through IS4 code in conjunction with my project. Hmmm, every second run so far has shown:

![image](https://user-images.githubusercontent.com/1901832/89171089-fd1b8680-d5c3-11ea-9753-f7c9269c42d1.png)

Still looking into this...

**Comment by kentcb** [comment]
OK, yeah, found this particular issue was due to overflow from this in `DefaultRefreshTokenService.cs`:

```C#
var newLifetime = currentLifetime + client.SlidingRefreshTokenLifetime;
```

So unfortunately, I'm back to square one trying to figure out the problem I am experiencing in production, which may or may not have anything to do with IS4. And my original question stands: what's the best way to have an indefinite refresh token? Should I just set the `AbsoluteRefreshTokenLifetime` to `int.MaxValue` (68 years, I believe)?
